File: 2019 KAKAO BLIND RECRUITMENT/길 찾기 게임.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Node {
    int id;
 	int x, y;   
    Node* leftChild;
    Node* rightChild;
};

bool cmp(Node* a, Node* b) {
    if(a->y == b->y) return a->x < b->x;
    return a->y > b->y;
}

void addNode(Node* root, Node* n) {
    if(n->x > root->x) {
        if(root->rightChild == NULL) root->rightChild = n;
        else addNode(root->rightChild, n);
    }
    if(n->x < root->x) {
        if(root->leftChild == NULL) root->leftChild = n;
        else addNode(root->leftChild, n);
    }
}

void preorder(Node* root, vector<int>& v) {
    if(root == NULL) return;
    v.push_back(root->id);
    preorder(root->leftChild, v);
    preorder(root->rightChild, v);
}

void postorder(Node* root, vector<int>& v) {
    if(root == NULL) return;
    postorder(root->leftChild, v);
    postorder(root->rightChild, v);
    v.push_back(root->id);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<Node*> nodes;
    for(int i = 0; i < nodeinfo.size(); ++i) {
        Node* node = new Node();
        node->id = i+1;
        node->x = nodeinfo[i][0];
        node->y = nodeinfo[i][1];
        nodes.push_back(node);
    }
    sort(nodes.begin(), nodes.end(), cmp);
    Node* root = nodes[0];
   	for(int i = 1; i < nodes.size(); ++i)
        addNode(root, nodes[i]);
    vector<int> pre, post;
    preorder(root, pre);
    postorder(root, post);
    vector<vector<int>> answer = { pre, post };
    return answer;
}

File: 2019 KAKAO BLIND RECRUITMENT/길 찾기 게임.java
import java.util.*;

class Solution {
    class Node {
        Node(int id, int x, int y) {
            this.id = id;
            this.x = x;
            this.y = y;
        }
        int id;
        int x, y;
        Node left;
        Node right;
    };
    int Idx;
    List<Node> Nodes = new ArrayList<Node>();
    
    Comparator<Node> Comp = new Comparator<Node>() {
        public int compare(Node a, Node b) {
            if (a.y == b.y)
                return a.x - b.x;
            
            return b.y - a.y;
        }
    };
    
    void addNode(Node parent, Node child) {
        if (child.x < parent.x) {
            if(parent.left == null)
                parent.left = child;
            else
                addNode(parent.left, child);
        } else {
            if(parent.right == null)
                parent.right = child;
            else
                addNode(parent.right, child);
        }
    }
    
    void preorder(int[][] answer, Node node) {
        if (node == null) return;
        
        answer[0][Idx++] = node.id;
        preorder(answer, node.left);
        preorder(answer, node.right);
    }
    
    void postorder(int[][] answer, Node node) {
        if (node == null) return;
        
        postorder(answer, node.left);
        postorder(answer, node.right);
        answer[1][Idx++] = node.id;
    }
    
    public int[][] solution(int[][] nodeinfo) {
        int size = nodeinfo.length;
        for (int i = 0; i < size; ++i) {
            Nodes.add(new Node(i+1, nodeinfo[i][0], nodeinfo[i][1]));
        }
        Nodes.sort(Comp); 
        
        Node root = Nodes.get(0);
        for (int i = 1; i < size; ++i)
            addNode(root, Nodes.get(i));
        
        int[][] answer = new int[2][size];
        preorder(answer, root);
        Idx = 0;
        postorder(answer, root);
        return answer;
    }
}
